lib/puppet/util/windows/file.rb

```diff
@@ -89,10 +89,31 @@ def symlink(target, symlink)
   end
   module_function :symlink
 
+
+  def file_exist(file_name)
+    path = file_name.to_s
+    attempts = 10
+    loop do
+      result = get_attributes(path,false)
+      # return false for file not found
+      return false if result == INVALID_FILE_ATTRIBUTES
+      # return true if we found a file/dir and it's not a symlink
+      return true if (result & FILE_ATTRIBUTE_REPARSE_POINT) != FILE_ATTRIBUTE_REPARSE_POINT
+      # walk the symlink and try again...
+      path = readlink(path)
+
+      return false if attempts <= 0
+      attempts -= 1
+    end
+  end
+  module_function :file_exist
+
+
   INVALID_FILE_ATTRIBUTES = 0xFFFFFFFF #define INVALID_FILE_ATTRIBUTES (DWORD (-1))
 
-  def get_attributes(file_name)
+  def get_attributes(file_name, raise_on_error = true)
     result = GetFileAttributesW(wide_string(file_name.to_s))
+    return result unless raise_on_error
     return result unless result == INVALID_FILE_ATTRIBUTES
     raise Puppet::Util::Windows::Error.new("GetFileAttributes(#{file_name})")
```

Comment:
I think the multiple returns / conditions makes things a little less clean - maybe better to refactor like this:

```ruby
def get_attributes(file_name, raise_on_invalid = true)
  result = GetFileAttributesW(wide_string(file_name.to_s))
  if raise_on_invalid && result == INVALID_FILE_ATTRIBUTES
    raise Puppet::Util::Windows::Error.new("GetFileAttributes(#{file_name})")
  end

  result
end
```

I'm still a little torn about whether we should file a ticket to make the default `false` for the raise in a future release. We can easily change the two other callsites at `add_attributes` (only used in 1 place in `spec/integration/util/windows/security_spec.rb`) and `remove_attributes` (only used in 1 place in `puppet/util/windows/security.rb`) to throw if necessary.  I would be surprised if we had any downstream consumers in module code, etc.